File: csharp/meetup/Meetup.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Schedule
{
    Teenth,
    First,
    Second,
    Third,
    Fourth,
    Last,
}

public static class ScheduleExtension
{
    public static Index GetIndex(this Schedule schedule) => schedule switch
    {
        Schedule.Teenth or Schedule.First => 0,
        Schedule.Second => 1,
        Schedule.Third => 2,
        Schedule.Fourth => 3,
        Schedule.Last => ^1,
        _ => throw new ArgumentOutOfRangeException(nameof(schedule)),
    };
} 

public class Meetup
{
    public int Month { get; }

    public int Year { get; }

    public Meetup(int month, int year)
    {
        Month = month;
        Year = year;
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule) 
        => datesInMonth(dayOfWeek, schedule).ElementAt(schedule.GetIndex());

    private IEnumerable<DateTime> datesInMonth(DayOfWeek dayOfWeek, Schedule schedule)
        => Enumerable.Range(1, DateTime.DaysInMonth(Year, Month))
        .Select(dayNumber => new DateTime(Year, Month, dayNumber))
        .Where(date => date.DayOfWeek == dayOfWeek)
        .SkipWhile(date => schedule == Schedule.Teenth && date.Day < 13);
}

File: go/meetup/meetup.go
// Package meetup is utility to find a suitable date.
package meetup

import (
	"time"
)

// WeekSchedule can be one of five week values: first, second, third, fourth, last, teenth.
type WeekSchedule int

const (
	Teenth WeekSchedule = iota - 2
	Last
	First
	Second
	Third
	Fourth
)

// Day returns the exact date of a meetup, given a month, year, weekday and week.
func Day(wSched WeekSchedule, wDay time.Weekday, month time.Month, year int) int {
	meetup := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
	dayslater := func(n int) {
		meetup = meetup.AddDate(0, 0, n)
	}
	for meetup.Weekday() != wDay {
		dayslater(1)
	}

	switch wSched {
	case Teenth:
		for meetup.Day() < 13 {
			dayslater(7)
		}
	case Last:
		dayslater(28)
		if meetup.Month() != month {
			dayslater(-7)
		}
	default:
		dayslater(int(wSched) * 7)
	}
	return meetup.Day()
}
